src/coreComponents/mpiCommunications/CommunicationTools.cpp

```diff
@@ -777,28 +803,97 @@ void CommunicationTools::synchronizePackSendRecv( const std::map< string, string
 }
 
-void CommunicationTools::synchronizeUnpack( MeshLevel & mesh,
-                                            std::vector< NeighborCommunicator > & neighbors,
-                                            MPI_iCommData & icomm,
-                                            bool on_device )
+
+bool CommunicationTools::asyncUnpack( MeshLevel & mesh,
+                                      std::vector< NeighborCommunicator > & neighbors,
+                                      MPI_iCommData & icomm,
+                                      bool onDevice,
+                                      parallelDeviceEvents & events )
 {
   GEOSX_MARK_FUNCTION;
 
-  // unpack the buffers
-  for( std::size_t count=0; count<neighbors.size(); ++count )
+  if( onDevice )
   {
-    int neighborIndex;
-    MpiWrapper::waitany( icomm.size,
+    size_t eventCount = 0;
+    for( const auto & sa : icomm.fieldNames )
+    {
+      eventCount += sa.second.size( );
+    }
+    eventCount *= neighbors.size( );
+    if( eventCount > events.size( ) )
+    {
+      events.reserve( eventCount );
+    }
+  }
+
+  int neighborIndex = 0;
+  int flag = 1;
+
+  while( flag && neighborIndex >= 0 )
+  {
+    flag = 0;
+    neighborIndex = -1;
+    MpiWrapper::testany( icomm.size,
                          icomm.mpiRecvBufferRequest.data(),
                          &neighborIndex,
+                         &flag,
                          icomm.mpiRecvBufferStatus.data() );
 
-    NeighborCommunicator & neighbor = neighbors[neighborIndex];
-    neighbor.unpackBufferForSync( icomm.fieldNames, mesh, icomm.commID, on_device );
+    // flag can true for MPI_REQUEST_NULLs from previous calls, so we also need to check that neighborIndex has been set
+    if( flag && neighborIndex >= 0 )
+    {
+      // unpack the recvd buffer
+      NeighborCommunicator & neighbor = neighbors[ neighborIndex ];
+      neighbor.unpackBufferForSync( icomm.fieldNames, mesh, icomm.commID, onDevice, events );
+    }
+  }
+
+  // we don't want to check if the request is complete,
+  //  we want to check that we've processed the resulting buffer
+  //  which means that we've testany-d the request and it has been
+  //  deallocated and set to MPI_REQUEST_NULL
+  int allDone = true;
+  const MPI_Request * reqs = icomm.mpiRecvBufferRequest.data( );
+  for( int idx = 0; idx < icomm.size; ++idx )
+  {
+    if( reqs[ idx ] != MPI_REQUEST_NULL )
+    {
+      allDone = false;
+      break;
+    }
+  }
+
+  return allDone;
+}
+
+void CommunicationTools::finalizeUnpack( MeshLevel & mesh,
+                                         std::vector< NeighborCommunicator > & neighbors,
+                                         MPI_iCommData & icomm,
+                                         bool onDevice,
+                                         parallelDeviceEvents & events )
+{
+  GEOSX_MARK_FUNCTION;
+
+  // wait to unpack any remaining buffers
+  while( !asyncUnpack( mesh, neighbors, icomm, onDevice, events ) )
```

Comment:
I'd add some comment here or some mention that we have a spinning loop here. If we were to say launch this function asynchronously it would be terrible because it would be wasting tons of cycles.

Reply:
Do we have other options besides a while loop? It seems like there is a while loop here, and inside `asyncUnpack` there is one as well. I always prefer for loops with some high bound and a `break` to `while` loops if there is any danger of spinning.

Reply:
I'd agree that a for-loop is probably better. Ideally we'd check and then sleep for some time before checking again, but that would involve some thought to pick a good amount of time to sleep for and a max timeout. It doesn't matter now, but it might in the future.